x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/InferenceWaitForAllocation.java

```diff
@@ -0,0 +1,197 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Predicate;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * Class for storing inference requests for ml trained models while
+ * scaling is in progress. Once the trained model has at least 1
+ * allocation the stored requests are forwarded to a consumer for
+ * processing.Requests will timeout while waiting for scale.
+ */
+public class InferenceWaitForAllocation {
+
+    public static final int MAX_PENDING_REQUEST_COUNT = 100;
+
+    /**
+     * Track details of the pending request
+     */
+    public record WaitingRequest(
+        InferModelAction.Request request,
+        InferModelAction.Response.Builder responseBuilder,
+        TaskId parentTaskId,
+        ActionListener<InferModelAction.Response> listener
+    ) {
+        public String deploymentId() {
+            return request.getId();
+        }
+    }
+
+    private static final Logger logger = LogManager.getLogger(InferenceWaitForAllocation.class);
+
+    private final TrainedModelAssignmentService assignmentService;
+    private final BiConsumer<WaitingRequest, TrainedModelAssignment> queuedConsumer;
+    private AtomicInteger pendingRequestCount = new AtomicInteger();
+
+    /**
+     * Create with consumer of the successful requests
+     * @param assignmentService            Trained model assignment service
+     * @param onInferenceScaledConsumer    The consumer of the waiting request called once an
+     *                                     allocation is available.
+     */
+    public InferenceWaitForAllocation(
+        TrainedModelAssignmentService assignmentService,
+        BiConsumer<WaitingRequest, TrainedModelAssignment> onInferenceScaledConsumer
+    ) {
+        this.assignmentService = assignmentService;
+        this.queuedConsumer = onInferenceScaledConsumer;
+    }
+
+    /**
+     * Wait for at least 1 allocation to be started then process the
+     * inference request.
+     * If the pending request count is greater than {@link #MAX_PENDING_REQUEST_COUNT}
+     * the request listener is failed with a too many requests exception
+     * The timeout is the inference request timeout.
+     * @param request The inference request details
+     */
+    public synchronized void waitForAssignment(WaitingRequest request) {
+        logger.info("waitForAssignment will wait for condition");
+        if (pendingRequestCount.get() > MAX_PENDING_REQUEST_COUNT) {
+            request.listener.onFailure(
+                new ElasticsearchStatusException(
+                    "Rejected inference request waiting for an allocation of deployment [{}]. Too many pending requests",
+                    RestStatus.TOO_MANY_REQUESTS,
+                    request.request.getId()
+                )
+            );
+            return;
+        }
+
+        pendingRequestCount.incrementAndGet();
```

Comment:
There's a race condition here that may lead to > MAX_PENDING_REQUEST_COUNT requests.

request 1: check pendingRequestCount > MAX_PENDING_REQUEST_COUNT => ok
request 2: check pendingRequestCount > MAX_PENDING_REQUEST_COUNT => ok
request 1: pendingRequestCount.incrementAndGet()
request 2: pendingRequestCount.incrementAndGet()

Doesn't really matter that much to have 101 in-flight requests though...

Reply:
Note: I think the proper way of doing this is:
```
 if (pendingRequestCount.incrementAndGet() > MAX_PENDING_REQUEST_COUNT) {
    pendingRequestCount.decrementAndGet();
    (...)
}
```